### ISSUE TITLE
PPO Network does not learn further

### ISSUE THREAD

**Issue by CesMak** [henrycharlesworth/big2_PPOalgorithm#9]
Hey Henry,

I finally was able to setup PPO with your hyperparamters in my pytorch environment for my custom game (witches). Yet I was not able to include all your great adjustments. Currently I just reward after each step(one card is played) and I just use the real rewards. I think this might be the reason why my ai does not learn further (than learning the correct moves...).

Here you see an example of the learning process:

![image](https://user-images.githubusercontent.com/20240718/77644021-abe9ef00-6f60-11ea-952c-0caebdfbe77a.png)
* If an invalid move is done the game is reset and -100 reward is given
* Second is the reward (max. is +5 for one game (=15 cards played))
* 3 Random players play against one ai player (orange)

If I play 1000 games the ai just wins 50% of all games.

I would be very delighted if you have a look on my code and share your ideas what you think is the problem, that my net is not learning further... see **Code snippet**

**Rewarding**
As you can see I just use the true real rewards to learn. 

In your paper you used:

* ![image](https://user-images.githubusercontent.com/20240718/77644652-9f19cb00-6f61-11ea-932b-8bd4f6c932e7.png)

* Does this mean I have to use the evaluate function to get the estimated value and then implement the same forumla as you do?

**Code snippet**
```

class Memory:
    def __init__(self):
        self.actions = []
        self.states = []
        self.logprobs = []
        self.rewards = []
        self.is_terminals = []

    def clear_memory(self):
        del self.actions[:]
        del self.states[:]
        del self.logprobs[:]
        del self.rewards[:]
        del self.is_terminals[:]

class ActorMod(nn.Module):
    def __init__(self, state_dim, action_dim, n_latent_var):
        super(ActorMod, self).__init__()
        self.l1      = nn.Linear(state_dim, n_latent_var)
        self.l1_tanh = nn.PReLU()
        self.l2      = nn.Linear(n_latent_var, n_latent_var)
        self.l2_tanh = nn.PReLU()
        self.l3      = nn.Linear(n_latent_var+60, action_dim)

    def forward(self, input):
        x = self.l1(input)
        x = self.l1_tanh(x)
        x = self.l2(x)
        out1 = self.l2_tanh(x) # 64x1
        if len(input.shape)==1:
            out2 = input[180:240]   # 60x1 this are the available options of the active player!
            output =torch.cat( [out1, out2], 0)
        else:
            out2 = input[:, 180:240]
            output =torch.cat( [out1, out2], 1) #how to do that?
        x = self.l3(output)
        return x.softmax(dim=-1)


class ActorCritic(nn.Module):
    def __init__(self, state_dim, action_dim, n_latent_var):
        super(ActorCritic, self).__init__()

        # actor
        #TODO see question: https://discuss.pytorch.org/t/pytorch-multiple-inputs-in-sequential/74040
        tmp = ActorMod(state_dim, action_dim, n_latent_var)
        self.action_layer = tmp

        # critic
        self.value_layer = nn.Sequential(
                nn.Linear(state_dim, n_latent_var),
                nn.PReLU(),#prelu
                nn.Linear(n_latent_var, n_latent_var),
                nn.PReLU(),
                nn.Linear(n_latent_var, 1)
                )

    def forward(self, state_input):
        return torch.tensor(self.act(state_input, None))

    def act(self, state, memory):
        if type(state) is np.ndarray:
            state = torch.from_numpy(state).float().to(device)
        action_probs = self.action_layer(state)
        # here make a filter for only possible actions!
        #action_probs = action_probs *state[120:180]
        dist = Categorical(action_probs)
        action = dist.sample()

        if memory is not None:
            memory.states.append(state)
            memory.actions.append(action)
            memory.logprobs.append(dist.log_prob(action))

        return action.item()

    def evaluate(self, state, action):
        action_probs = self.action_layer(state)
        dist = Categorical(action_probs)

        action_logprobs = dist.log_prob(action)
        dist_entropy = dist.entropy()

        state_value = self.value_layer(state)

        return action_logprobs, torch.squeeze(state_value), dist_entropy

class PPO:
    def __init__(self, state_dim, action_dim, n_latent_var, lr, betas, gamma, K_epochs, eps_clip):
        self.lr = lr
        self.betas = betas
        self.gamma = gamma
        self.eps_clip = eps_clip
        self.K_epochs = K_epochs

        self.policy = ActorCritic(state_dim, action_dim, n_latent_var).to(device)
        self.optimizer = torch.optim.Adam(self.policy.parameters(), lr=lr, betas=betas,  eps=1e-5) # no eps before!
        self.policy_old = ActorCritic(state_dim, action_dim, n_latent_var).to(device)
        self.policy_old.load_state_dict(self.policy.state_dict())
        #TO decay learning rate during training:
        self.scheduler = torch.optim.lr_scheduler.StepLR(self.optimizer, step_size=20000, gamma=0.5)
        self.MseLoss = nn.MSELoss()

    def update(self, memory):
        # What to do with the rewards???
        # Normalizing the rewards:
        rewards = torch.tensor(memory.rewards).to(device)                     # use here memory.rewards
        #rewards = (rewards - rewards.mean()) / (rewards.std() + 1e-5)  # commented out
        rewards = rewards/100

        # convert list to tensor
        old_states = torch.stack(memory.states).to(device).detach()
        old_actions = torch.stack(memory.actions).to(device).detach()
        old_logprobs = torch.stack(memory.logprobs).to(device).detach()

        # Optimize policy for K epochs:
        for _ in range(self.K_epochs):
            # Evaluating old actions and values :
            logprobs, state_values, dist_entropy = self.policy.evaluate(old_states, old_actions)

            # Finding the ratio (pi_theta / pi_theta__old):
            ratios = torch.exp(logprobs - old_logprobs.detach())

            # Finding Surrogate Loss:
            advantages = rewards - state_values.detach()
            surr1 = ratios * advantages
            surr2 = torch.clamp(ratios, 1-self.eps_clip, 1+self.eps_clip) * advantages
            #0.5 is alpha1 in Big2 paper
            #0.01 is alpha2 in big2 paper (used 0.02)
            # IS MseLoss the squared error loss?
            loss = -torch.min(surr1, surr2) + 0.5*self.MseLoss(state_values, rewards) - 0.02*dist_entropy

            # take gradient step
            self.optimizer.zero_grad()
            loss.mean().backward()
            self.scheduler.step()#sheduler to lower the learning rate !
            self.optimizer.step()

        # Copy new weights into old policy:
        self.policy_old.load_state_dict(self.policy.state_dict())
        return rewards.mean(), len(rewards)


def main():
    start_time = datetime.datetime.now()
    #stdout.write_file("hallo.txt")
    ############## Hyperparameters ##############
    env_name = "Witches-v0"
    # creating environment
    env = gym.make(env_name)
    state_dim = env.observation_space.n
    action_dim = env.action_space.n

    log_path      = "logging.txt"
    try:
        os.remove(os.getcwd()+"/"+log_path)
    except:
        print("No Logging to be removed!")
    render        = False
    solved_reward = 230         # stop training if avg_reward > solved_reward
    log_interval  = 50           # print avg reward in the interval
    max_episodes  = 500000       # max training episodes
    # TODO DO NOT RESET AFTER FIXED VALUE BUT AT END OF Game
    # THIS DEPENDS IF YOU DO ALLOW TO LEARN THE RULES!
    n_latent_var    = 512            # number of variables in hidden layer
    update_timestep = 5             # before 5 in big2 = 5
    lr              =  0.00025      # in big2game:  0.00025
    gamma           = 0.99
    betas           = (0.9, 0.999)
    K_epochs        = 4               # update policy for K epochs in big2game:nOptEpochs = 5  typical 3 - 10 is the number of passes through the experience buffer during gradient descent.
    eps_clip        = 0.2             # clip parameter for PPO Setting this value small will result in more stable updates, but will also slow the training process.
    random_seed     = None
    decay           = 50000
    #############################################

    if random_seed:
        torch.manual_seed(random_seed)
        env.seed(random_seed)

    memory = Memory()
    ppo = PPO(state_dim, action_dim, n_latent_var, lr, betas, gamma, K_epochs, eps_clip)
    print("Learning Parameters:", lr, n_latent_var, "Update_timestep", update_timestep, K_epochs, eps_clip, betas)

    # logging variables
    running_reward = 0
    avg_length = 0
    timestep = 0
    reward_mean = 0
    wrong_moves = 0
    invalid_moves = 0
    total_number_of_games_played = 0
    total_rewards  = 0
    total_games_won = np.zeros(4,)

    # training loop
    for i_episode in range(1, max_episodes+1):
        timestep += 1
        state = env.reset()
        done  = 0
        while not done:
            # Running policy_old:
            # state has to be right before the AI Plays!
            action = ppo.policy_old.act(state, memory)

            # this should be the reward for the above action
            # this is the new state! when the ai player is again
            state, reward, done, nu_games_won = env.step(action)
            if reward==-100:
                invalid_moves +=1

            total_rewards += reward
            memory.rewards.append(reward)
            memory.is_terminals.append(done)

        total_number_of_games_played+=1
        total_games_won +=nu_games_won

        # update if its time
        if timestep % update_timestep == 0:
            reward_mean, wrong_moves = ppo.update(memory)
            memory.clear_memory()
            timestep = 0

        # logging
        if i_episode % decay == 0:
            ppo.eps_clip *=0.4
            print(ppo.eps_clip)

        if i_episode % log_interval == 0:
            total_reward_per_game_positive = total_rewards/log_interval
            per_game_reward = total_reward_per_game_positive-15*21
            games_won = str(np.array2string(total_games_won))
            # total_rewards per game should be maximized!!!!
            aaa = ('Game ,{:07d}, reward ,{:0.5}, invalid_moves ,{:4.4}, games_won ,{},  Time ,{},\n'.format(total_number_of_games_played, per_game_reward, invalid_moves/log_interval, games_won, datetime.datetime.now()-start_time))
            print(aaa)
            if per_game_reward>-2.5:
                 #print("Export ONNX:::")
                 torch.save(ppo.policy.state_dict(), './PPO_{}_{}_{}.pth'.format(env_name, per_game_reward, total_games_won[1]))
                 #exportONNX(ppo.policy, torch.rand(240), str(per_game_reward))
            invalid_moves = 0
            total_rewards = 0
            total_games_won = np.zeros(4,)
            with open(log_path, "a") as myfile:
                myfile.write(aaa)

if __name__ == '__main__':
main()
```

**Comment by henrycharlesworth** [comment]
Hi,

I haven't had time to go through your code in detail, but when you say you use "the real rewards to learn" I don't think this is the right thing to do. In reinforcement learning, the aim is to maximise long-term (discounted) reward, i.e. \sum_{t=0}^{\infty} \gamma^t r_t - here \gamma is the discounting factor, and so in PPO (and other actor-critic policy gradient methods) you calculate an estimate for the advantage of the action in terms of the RETURN (discounted sum of future rewards) rather than just the reward received at the next step - because maximising this would be extremely short-sighted (and is not what RL is about, anyway). The formula in that image is the formula for generalized advantage estimation (https://arxiv.org/abs/1506.02438) that is usually used with PPO. There are other ways to estimate the advantage, but what you do:
`advantages = rewards - state_values.detach()`
is not right. I would recommend looking at this PPO implementation (in Pytorch): https://github.com/ikostrikov/pytorch-a2c-ppo-acktr-gail
It has both generalised advantage estimation and an alternative simpler way to estimate the return (in this file: https://github.com/ikostrikov/pytorch-a2c-ppo-acktr-gail/blob/master/a2c_ppo_acktr/storage.py)

**Comment by CesMak** [comment]
Hey,

thx. for sharing **a2c-ppo-acktr-gail**. I already saw that repro, however I wanted to start with this [ppo_minmal](https://github.com/nikhilbarhate99/PPO-PyTorch) implementation. It turns out that instead of GAE it uses so called monte carlo state estimation rewarding:
( I also did not found it in a paper yet):

```python
    def monteCarloRewards(self, memory):
        # Monte Carlo estimate of state rewards:
        # see: https://medium.com/@zsalloum/monte-carlo-in-reinforcement-learning-the-easy-way-564c53010511
        rewards = []
        discounted_reward = 0
        for reward, is_terminal in zip(reversed(memory.rewards), reversed(memory.is_terminals)):
            if is_terminal:
                discounted_reward = 0
            discounted_reward = reward + (self.gamma * discounted_reward)
            rewards.insert(0, discounted_reward)

        # Normalizing the rewards:
        rewards = torch.tensor(rewards).to(device)                     # use here memory.rewards
        rewards = (rewards - rewards.mean()) / (rewards.std() + 1e-5)  # commented out
        return rewards

# used in: 
    def monteCarloRewards(self, memory):
        # Monte Carlo estimate of state rewards:
        # see: https://medium.com/@zsalloum/monte-carlo-in-reinforcement-learning-the-easy-way-564c53010511
        rewards = []
        discounted_reward = 0
        for reward, is_terminal in zip(reversed(memory.rewards), reversed(memory.is_terminals)):
            if is_terminal:
                discounted_reward = 0
            discounted_reward = reward + (self.gamma * discounted_reward)
            rewards.insert(0, discounted_reward)

        # Normalizing the rewards:
        rewards = torch.tensor(rewards).to(device)                     # use here memory.rewards
        rewards = (rewards - rewards.mean()) / (rewards.std() + 1e-5)  # commented out
        return rewards
```

With correct hyperparameter tuning I was able to achieve a 81% win rate of the ai. Which is kind of nice but I think there is a lot more.... 

I tried to include gae reward estimation with this function:
```python
    def get_advantages(self, values, masks, rewards, gamma):
        returns = []
        gae = 0
        for i in reversed(range(len(rewards))):
            delta = rewards[i-1] + gamma * values[i] * masks[i-1] - values[i-1]
            gae = delta + gamma * 0.95 * masks[i-1] * gae
            returns.insert(0, gae + values[i-1])

        adv = np.array(returns) - values.detach().numpy()
        adv = torch.tensor(adv.astype(np.float32)).float()
        # Normalizing advantages
        return returns, (adv - adv.mean()) / (adv.std() + 1e-5)
```
it did not work yet. I am not exactly sure what values I have to use for that function and also how to get the next value.
